client/pluginmanager/drivermanager/manager.go

```diff
@@ -0,0 +1,320 @@
+	instances   map[string]*instanceManager
+	instancesMu sync.Mutex
+
+	// reattachConfigs stores the plugin reattach configs
+	reattachConfigs    map[loader.PluginID]*shared.ReattachConfig
+	reattachConfigLock sync.Mutex
+
+	// allows/block lists
+	allowedDrivers map[string]struct{}
+	blockedDrivers map[string]struct{}
+
+	// readyCh is ticked once at the end of Run()
+	readyCh chan struct{}
+}
+
+// New returns a new driver manager
+func New(c *Config) *manager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &manager{
+		logger:          c.Logger.Named("driver_mgr"),
+		state:           c.State,
+		ctx:             ctx,
+		cancel:          cancel,
+		loader:          c.Loader,
+		pluginConfig:    c.PluginConfig,
+		updater:         c.Updater,
+		instances:       make(map[string]*instanceManager),
+		reattachConfigs: make(map[loader.PluginID]*shared.ReattachConfig),
+		allowedDrivers:  c.AllowedDrivers,
+		blockedDrivers:  c.BlockedDrivers,
+		readyCh:         make(chan struct{}),
+	}
+}
+
+// PluginType returns the type of plugin this mananger mananges
+func (*manager) PluginType() string { return base.PluginTypeDriver }
+
+// Run starts the mananger, initializes driver plugins and blocks until Shutdown
+// is called.
+func (m *manager) Run() {
+	// Load any previous plugin reattach configuration
+	m.loadReattachConfigs()
+
+	// Get driver plugins
+	driversPlugins := m.loader.Catalog()[base.PluginTypeDriver]
+	if len(driversPlugins) == 0 {
+		m.logger.Debug("exiting since there are no driver plugins")
+		m.cancel()
+		return
+	}
+
+	var skippedDrivers []string
+	for _, d := range driversPlugins {
+		id := loader.PluginInfoID(d)
+		// Skip drivers that are not in the allowed list if it is set.
+		if _, ok := m.allowedDrivers[id.Name]; len(m.allowedDrivers) > 0 && !ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+		// Skip fingerprinting drivers that are in the blocked list
+		if _, ok := m.blockedDrivers[id.Name]; ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+
+		storeFn := func(c *plugin.ReattachConfig) error {
+			return m.storePluginReattachConfig(id, c)
+		}
+		fetchFn := func() (*plugin.ReattachConfig, bool) {
+			return m.fetchPluginReattachConfig(id)
+		}
+
+		instance := newInstanceManager(&instanceManagerConfig{
+			Logger:               m.logger,
+			Ctx:                  m.ctx,
+			Loader:               m.loader,
+			StoreReattach:        storeFn,
+			FetchReattach:        fetchFn,
+			PluginConfig:         m.pluginConfig,
+			ID:                   &id,
+			UpdateNodeFromDriver: m.updater,
+		})
+
+		m.instancesMu.Lock()
+		m.instances[id.Name] = instance
+		m.instancesMu.Unlock()
+	}
+
+	if len(skippedDrivers) > 0 {
+		m.logger.Debug("drivers skipped due to allow/block list", "skipped_drivers", skippedDrivers)
+	}
+
+	// signal ready
+	select {
+	case <-m.ctx.Done():
+		return
+	case m.readyCh <- struct{}{}:
+	}
+
+	// wait for shutdown
+	<-m.ctx.Done()
```

Comment:
The contract I created in the interface was that `Run()` would block until `Shutdown()` was called (and had returned). There's no reason as to why, other than to be consistent one way or the other. It would require changing the logging in plugin group: https://github.com/hashicorp/nomad/blob/2a810d0a6e2da4b1cc61dd1afa46dbe45823f833/client/pluginmanager/group.go#L50-L54